substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVTypeRecord.java

```diff
@@ -0,0 +1,234 @@
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null || obj.getClass() != this.getClass()) {
+            return false;
+        }
+        return this.type == ((CVTypeRecord)obj).type;
+    }
+
+    @Override
+    public String toString() {
+        return String.format("CVTypeRecord seq=0x%04x type=0x%04x pos=0x%04x ", sequenceNumber, type, startPosition);
+    }
+
+    private static int alignPadded4(byte[] buffer, int originalpos) {
+        int pos = originalpos;
+        int align = pos & 3;
+        if (align == 1) {
+            byte[] p3 = {LF_PAD3, LF_PAD2, LF_PAD1};
+            pos = CVUtil.putBytes(p3, buffer, pos);
+        } else if (align == 2) {
+            pos = CVUtil.putByte(LF_PAD2, buffer, pos);
+            pos = CVUtil.putByte(LF_PAD1, buffer, pos);
+        } else if (align == 3) {
+            pos = CVUtil.putByte(LF_PAD1, buffer, pos);
+        }
+        return pos;
+    }
+
+    static final class CVTypeProcedureRecord extends CVTypeRecord {
+
+        int returnType = -1;
+        CVTypeArglistRecord argList = null;
+
+        CVTypeProcedureRecord() {
+            super(LF_PROCEDURE);
+        }
+
+        public CVTypeProcedureRecord returnType(int leaf) {
+            this.returnType = leaf;
+            return this;
+        }
+
+        public CVTypeProcedureRecord returnType(CVTypeRecord leaf) {
+            this.returnType = leaf.getSequenceNumber();
+            return this;
+        }
+
+        CVTypeProcedureRecord argList(CVTypeArglistRecord leaf) {
+            this.argList = leaf;
+            return this;
+        }
+
+        @Override
+        public int computeSize(int initialPos) {
+            return computeContents(null, initialPos);
+        }
+
+        @Override
+        public int computeContents(byte[] buffer, int initialPos) {
+            int pos = CVUtil.putInt(returnType, buffer, initialPos);
+            pos = CVUtil.putByte((byte) 0, buffer, pos); /* callType */
+            pos = CVUtil.putByte((byte) 0, buffer, pos); /* funcAttr */
+            pos = CVUtil.putShort((short) argList.getSize(), buffer, pos);
+            pos = CVUtil.putInt(argList.getSequenceNumber(), buffer, pos);
+            return pos;
+        }
+
+        @Override
+        public String toString() {
+            return String.format("LF_PROCEDURE 0x%04x ret=0x%04x arg=0x%04x ", getSequenceNumber(), returnType, argList.getSequenceNumber());
+        }
+
+        @Override
+        public int hashCode() {
+            int h = type;
+            h = 31 * h + returnType;
+            h = 31 * h + argList.hashCode();
+            /* callType and funcAttr are always zero so do not add them to the hash */
+            return h;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            CVTypeProcedureRecord other = (CVTypeProcedureRecord)obj;
```

Comment:
I don't think this cast is safe without first checking if `super.equals(obj)` is `true`.

Reply:
Done.